<div align="center">
МІНІСТЕРСТВО ОСВІТИ І НАУКИ УКРАЇНИ<br>
ІНСТИТУТ ПРИКЛАДНОГО СИСТЕМНОГО АНАЛІЗУ<br>
НАЦІОНАЛЬНОГО ТЕХНІЧНОГО УНІВЕРСИТЕТУ УКРАЇНИ<br>
«КИЇВСЬКИЙ ПОЛІТЕХНІЧНИЙ ІНСТИТУТ»<br>
КАФЕДРА МАТЕМАТИЧНИХ МЕТОДІВ СИСТЕМНОГО АНАЛІЗУ<br>
</div>
<br><br><br><br><br><br>
<center><font size="+2">Лабораторна робота №1</font></center>
<center>З курсу "Чисельні методи"</center>
<center>Тема "Обчислення власних чисел та відповідих власних векторів матриць"</center>
<br><br><br>
<div align="right">
    <strong>Виконав:</strong><br>
    Студент группи КА-81<br>
    Єрко А.В.<br><br>
    <strong>Прийняла:</strong><br>
    Хоменко О.В.
</div>
<br><br><br><br><br><br>
<center>Київ - 2020</center>


<hr>

**Мета роботи**: набути вміння та досвід використання методів
розв’язання часткової та повної проблеми власних значень. Закріпити
та поглибити теоретичні знання методів розв’язання часткової та
повної проблеми власних значень.

In [1]:
import numpy as np #підключимо бібліотеку матричних розрахунків
np.set_printoptions(precision=6)

# Варіант 4
### Матриця з таблиці 1:

In [2]:
matrix1 = np.array([
    [ 932,  82,  -578, -519,  593],
    [  82,  834, -350, -100,  374],
    [-578, -350,  834,  352, -570],
    [-519, -100,  352,  653, -253],
    [ 593,  374, -570, -253,  898]               
    ])

### Матриця з таблиці 2:

In [3]:
matrix2 = np.array([
    [ 863, -758, -380, -208, -294],
    [ -17,  756, -345,  553,  771],
    [-856,  129,  872,  661,  654],
    [ 854,  818,  401,  878, -108],
    [-847,   78,  579,  599,  834]            
    ])

# Завдання 1 
Знайти найбільше та власне число та відповідний власний
вектор степеневим методом або методом скалярних добутків з точністю
$\varepsilon=0,0001$ (матриця з таблиці 1).

1. Звіт має містити код програми.
2. Для методу скалярних добутків на кожному кроці виводити $\lambda^{(k)} = \dfrac{s^{(k)}}{t^{(k)}}$

### Метод скалярних добутків

In [4]:
eps = 0.0001
A = np.copy(matrix1)
y = np.random.rand(5) #в якості першого наближення беремо випадковий вектор
print("First approximation ", y)

iteration = 0 #номер поточної ітерації
lmbd = ( A.dot(y) @ y ) / (y @ y) #де @ - скалярний добуток, dot - матричне множення
lmbd_prev = np.inf

while (np.abs(lmbd - lmbd_prev) >= eps):
    y = A.dot(y)
    lmbd_prev = lmbd
    lmbd = ( A.dot(y) @ y ) / (y @ y)
    iteration +=1 
    print(" > Iteration %2d, 𝜆 = %f" % (iteration, lmbd))

print("Approximation of major eigenvalue: ", lmbd)
y = A.dot(y)
x = y / np.linalg.norm(y)
print("Approximation of major eigenvector: ", x)

print("\nResults using numpy library:")
value, vect = np.linalg.eigh(A)
print(value[-1],vect[:,-1])

First approximation  [0.735692 0.450337 0.386997 0.823565 0.108041]
 > Iteration  1, 𝜆 = 1040.847953
 > Iteration  2, 𝜆 = 2140.825456
 > Iteration  3, 𝜆 = 2403.536504
 > Iteration  4, 𝜆 = 2437.082853
 > Iteration  5, 𝜆 = 2441.181873
 > Iteration  6, 𝜆 = 2441.688514
 > Iteration  7, 𝜆 = 2441.751486
 > Iteration  8, 𝜆 = 2441.759327
 > Iteration  9, 𝜆 = 2441.760304
 > Iteration 10, 𝜆 = 2441.760425
 > Iteration 11, 𝜆 = 2441.760440
Approximation of major eigenvalue:  2441.7604403527166
Approximation of major eigenvector:  [ 0.529928  0.278915 -0.508796 -0.342322  0.515097]

Results using numpy library:
2441.760442510218 [-0.529934 -0.278904  0.508795  0.342327 -0.515095]


Для підтвердження коректності обрахуємо вектор нев'язки

In [5]:
r = A.dot(x) - lmbd*x
print("Residual vector:", r)
print("Norm of residual vector:", np.linalg.norm(r))

Residual vector: [ 0.009039 -0.016499  0.001618 -0.007391 -0.003676]
Norm of residual vector: 0.020607629480361104


#### **Висновки:**
При заданій точності $\varepsilon = 10^{-4}$ для данної симетричної матриці метод скалярних добутків збігається доволі швидко, 8-12 ітерацій в залежності від першого наближення. Оскільки матриця симетрична, отримані результати були додатково перевірені за допомогою функції з бібліотеки ```numpy``` $-$ різниця між власним числами, отриманим методом скалярних добутків та функцією ```np.linalg.eigh``` такох не перевищує $\varepsilon$.

При точності $\varepsilon$ для власного числа, норма нев'язки для відповідного власного вектора вийшла значно більшою $-$ порядка $5 \cdot 10^{-2}$. Отже, якщо першочерговою задачею є знаходження саме власного вектора, доцільно брати за умову виходу з циклу зменшення норми нев'язки $\|Ax - \lambda x\| < \varepsilon$.

---
# Завдання 2
Розв'язати повну проблему власних значень:
1. методом Якобі  (матриця з таблиці 1);
2. QR або LR методом (матриця з таблиці 2);

#### Вимоги до звіту (в електронному вигляді) для другого завдання.
1. Звіт має містити код програми.
2. Для підтведження коректності роботи методу Якобі виведіть  на кожній ітерації такі величини:
    * матриця, що діагоналізується (А);
    * $i,j$ - індекси ключового елементу (максимального за модулем); кутові параметри $c, s$; при цьому не забудьте перевірити, що $c^2 + s^2 = 1$;
    * суму квадратів діагональних елементів $\delta$, недіагональних елементів $2\omega$, всіх елементів ( $\delta + 2\omega = \mathrm{const}$!). Тобто, перевірити, що норма Фробеніуса тобто $\|A\|_F = \sum\limits^n_{i=1} \sum\limits^n_{j=1} a_{ij}^2$ не змінюється, причому $\delta = \sum\limits^n_{i=1} a_{ii}^2$ зростає на $2a_{ij}^2$, а $2\omega = 2 \sum\limits^{n-1}_{i=1} \sum\limits^n_{j=i+1} a_{ij}^2$ зменшується на $2a_{ij}^2$.
    
3. Для методів QR та LR виводьте матриці Q, R або L, R, відповідно, на кожній ітерації. 

Для кожної власної пари $(\lambda, v)$ як підтвердження коректності результату вивести на друк значення вектору нев'язки $x = Av – \lambda v$, де $v$ – власний вектор, $\lambda$ - власне число. Точність $ \varepsilon= 10^{-4}$.

У висновках проаналізувати одержані результати.


## 2.1 Метод Якобі

In [6]:
def upper_nondiag(B):
    """
    (Допоміжна функція)
    Ця функція повертає матрицю, в якій елементи над головною діагоналлю
    дорівнюють відповідним елементам вихідної матриці, а усі інші - 0
    """
    i_idx, j_idx = np.mgrid[0:B.shape[0], 0:B.shape[1]]
    M = np.copy(B)
    M[i_idx>=j_idx] = 0 #розглядаємо лише елементи над головною діагоналлю
    return M

def max_nondiag(B):
    """
    Ця функція повертає індекс найбільшого за модулем недіагонального елемента
    симетричної матриці (B)
    """
    M = np.abs(upper_nondiag(B)) #матриця модулів елементів верхнього трикутника
    return np.unravel_index(np.argmax(M), M.shape)

def sqrsum_nondiag(B):
    """
    Ця функція повертає сумму квадратів верхніх недіагональних елементів матриці.
    """
    M = upper_nondiag(B)**2 #піднесення кожного елементу в квадрат
    return M.sum()

def sign(a):
    """
    Модифікована функція сігнум.
    Відрізняється тим, що повертає 1 при а==0, тим самим дозволяючи 
    не втратити загальність в основному циклі програми.
    """
    if a>=0: return 1
    return -1

In [7]:
eps = 0.0001
iteration = 0
B = np.copy(matrix1)
V = np.eye(B.shape[0]) #матриця, що в кінці буде містити власні вектори

while 2*sqrsum_nondiag(B) >= eps:
    iteration += 1
    print("\n>>> Iteration %d"%iteration)
    print("",B)
                       
    i, j = max_nondiag(B)
    print("Argmax (%d,%d)"%(i,j))

    tau = (B[i,i] - B[j,j]) / (2 * B[i,j])
    t = -tau + sign(tau)*np.sqrt(tau**2 + 1) 
    
    c = 1 / np.sqrt(1 + t**2)
    s = c*t
    print("c =% .10f, s =% .10f, c²+s² = %f" % (c,s,c**2 + s**2))
    
    delta = (np.diag(B)**2).sum()
    omega = 2*sqrsum_nondiag(B)
    print("𝛿 = %f, 2𝜔 = %f, sum = %f" % (delta,omega,delta+omega))
    
    T = np.eye(B.shape[0]) #будуємо матрицю плоского обертання
    T[i,i] = T[j,j] = c
    T[i,j] = -s
    T[j,i] = s
    
    V = V @ T        # @ в данному випадку інтерпретується як матричне множення
    B = T.T @ B @ T  # .T - операція транспонування

print("\n>>> RESULTS")
print("Matrix B =")
print(B)
print("Eigenvalues:")
print(np.diag(B))
print("Eigenvectors:")
print(V)

print("\nResults using numpy library:")
value, vect = np.linalg.eigh(matrix1)
print(value[::-1])
print(vect[:,::-1])


>>> Iteration 1
 [[ 932   82 -578 -519  593]
 [  82  834 -350 -100  374]
 [-578 -350  834  352 -570]
 [-519 -100  352  653 -253]
 [ 593  374 -570 -253  898]]
Argmax (0,4)
c = 0.7171666536, s = 0.6969017083, c²+s² = 1.000000
𝛿 = 3492549.000000, 2𝜔 = 3494014.000000, sum = 6986563.000000

>>> Iteration 2
 [[ 1.508244e+03  3.194489e+02 -8.117563e+02 -5.485256e+02  1.136868e-13]
 [ 3.194489e+02  8.340000e+02 -3.500000e+02 -1.000000e+02  2.110744e+02]
 [-8.117563e+02 -3.500000e+02  8.340000e+02  3.520000e+02 -5.975805e+00]
 [-5.485256e+02 -1.000000e+02  3.520000e+02  6.530000e+02  1.802488e+02]
 [ 5.684342e-14  2.110744e+02 -5.975805e+00  1.802488e+02  3.217564e+02]]
Argmax (0,2)
c = 0.8317268546, s =-0.5551850497, c²+s² = 1.000000
𝛿 = 4195847.000000, 2𝜔 = 2790716.000000, sum = 6986563.000000

>>> Iteration 3
 [[2050.098153  460.009       0.       -651.648631    3.317678]
 [ 460.009     834.       -113.751143 -100.        211.074388]
 [   0.       -113.751143  292.145473  -11.765374   -4.97

Для підтвердження коректності обрахуємо вектори нев'язки

In [8]:
A = np.copy(matrix1)
R = A@V - np.diag(B)*V #матриця векторів нев'язки

for index, lmbd in enumerate(np.diag(B)):
    print(">>> %d EIGENVALUE" % (index+1))
    print("(",lmbd,",",V[:,index],")")
    print("Residual:", R[:,index])
    print("Norm of residual vector:", np.linalg.norm(R[:,index]))
    print()

print("Norm of residual MATRIX:", np.linalg.norm(R))

>>> 1 EIGENVALUE
( 2441.7604425100776 , [ 0.529933  0.278904 -0.508795 -0.342327  0.515095] )
Residual: [ 1.554968e-04 -3.943763e-04  2.659204e-05 -1.994687e-04 -5.273450e-05]
Norm of residual vector: 0.000472215758385825

>>> 2 EIGENVALUE
( 861.8710772983119 , [-0.438611  0.80085  -0.078574  0.358277  0.17811 ] )
Residual: [-0.000232 -0.000112  0.000322  0.000136 -0.0002  ]
Norm of residual vector: 0.0004774461492828368

>>> 3 EIGENVALUE
( 271.06019712799525 , [ 0.159073  0.188505  0.845044 -0.243654  0.407056] )
Residual: [-4.294106e-05  7.638459e-05 -7.876866e-06  4.122597e-05  2.243671e-05]
Norm of residual vector: 9.971746338230464e-05

>>> 4 EIGENVALUE
( 433.0816034888191 , [ 0.07858  -0.400634 -0.008889  0.696408  0.59013 ] )
Residual: [ 0.002206  0.000914  0.000458  0.001434 -0.001359]
Norm of residual vector: 0.0031332289550770266

>>> 5 EIGENVALUE
( 143.22667957479646 , [-0.703784 -0.291221 -0.144134 -0.458364  0.434748] )
Residual: [-2.092112e-04  1.275068e-03 -7.857263e-06 

#### **Висновки**:
При даних умовах метод Якобі збігається за 24 ітерації, що є доволі хорошим результатом, враховуючи він дає усі власні значення та відповідні вектори, при цьому в випадку матриці $5\times5$ недіагональних елементів 10 пар, тобто в процессі досяжності рівня точності $\varepsilon = 10^{-4}$ кожна пара мінімізується в середньому 2-3 рази. Перевірки показали, що на кожній ітерації сума квадратів кутових коефіцієнтів дорівнює одиниці, а норма Фробеніуса матриці не змінюється (дорівнює 6986563) з великим рівнем точності. Додаткова перевірка за допомогою бібліотеки ```numpy```, результати якої, враховуючи невеликий заданий рівень точності $\varepsilon = 10^{-4}$, можемо прийняти за дійсні значення, показала що отримані методом Якобі власні числа та координати матриці власних векторів відрізняються від них не більш ніж на $\varepsilon$.

Для перевірки отриманих власних чисел та векторів, були обраховані вектори нев'язки для кожної пари. Норма виявилась порядку $5 \times 10^{-3}$ для найбільшого вектора нев'язки, що не є дуже гарним результатом, але, принаймні, кращим аніж в методі скалярних добутків.

## 2.2 QR метод

In [9]:
A = np.copy(matrix2)

from scipy.linalg import qr

iteration = 0
while sqrsum_nondiag(A.T) >= eps: #нижній трикутник = верхній транспонованої
    iteration += 1
    print("\n>>> Iteration %d"%iteration)
    print("A=")
    print(A)
    
    Q, R = qr(A)
    print("Q=")
    print(Q)
    print("R=")
    print(R)
    A = R @ Q
    
print("\n>>> RESULTS")
print("Matrix A = ")
print(A)
print("Eigenvalues:")
print(np.diag(A))

print("\nResults using numpy library:")
value, vect = np.linalg.eig(matrix2)
print(value)


>>> Iteration 1
A=
[[ 863 -758 -380 -208 -294]
 [ -17  756 -345  553  771]
 [-856  129  872  661  654]
 [ 854  818  401  878 -108]
 [-847   78  579  599  834]]
Q=
[[-0.504642  0.528606 -0.164039  0.652606 -0.114497]
 [ 0.009941 -0.558112  0.567288  0.563471 -0.221581]
 [ 0.500549 -0.063989 -0.520722  0.191977 -0.661324]
 [-0.499379 -0.635833 -0.572312  0.009008  0.136797]
 [ 0.495286 -0.026626 -0.22931   0.46869   0.694067]]
R=
[[-1710.12251     84.744806   711.332663   299.54696    950.389806]
 [    0.       -1353.058505  -334.506047 -1035.093417  -581.099337]
 [    0.           0.        -949.717088  -636.214157    15.618919]
 [    0.           0.           0.         591.408306   758.037208]
 [    0.           0.           0.           0.          -5.605164]]

>>> Iteration 2
A=
[[ 1.541027e+03 -1.212562e+03 -4.311745e+02 -4.835887e+02  4.072155e+02]
 [ 4.820676e+01  1.450182e+03  1.322599e+02 -1.108306e+03 -2.389105e+01]
 [-1.499318e+02  4.648819e+02  8.550700e+02 -1.807343e+02  5

Q=
[[-1.000000e+00 -1.836650e-04 -6.939577e-10 -2.795440e-26 -6.435835e-36]
 [ 1.836650e-04 -1.000000e+00 -9.475606e-07 -9.185161e-23 -1.597364e-32]
 [ 5.199240e-10  9.475608e-07 -1.000000e+00 -1.132360e-16 -7.340045e-27]
 [-4.778965e-26 -1.544640e-23  1.132360e-16 -1.000000e+00 -7.866721e-11]
 [-3.445250e-36 -1.023363e-32  1.567917e-27 -7.866721e-11  1.000000e+00]]
R=
[[-2069.094934  1050.73828   -312.922812   451.009145   311.797499]
 [    0.       -1293.728024  -605.66244    889.02392     98.024376]
 [    0.           0.        -698.471175  -407.310748   212.597224]
 [    0.           0.           0.        -104.078115   823.609031]
 [    0.           0.           0.           0.          37.4354  ]]

>>> Iteration 22
A=
[[ 2.069288e+03 -1.050359e+03  3.129218e+02 -4.510091e+02  3.117975e+02]
 [-2.376128e-01  1.293727e+03  6.056637e+02 -8.890239e+02  9.802438e+01]
 [-3.631519e-07 -6.618439e-04  6.984712e+02  4.073107e+02  2.125972e+02]
 [ 4.973857e-24  1.607633e-21 -1.178539e-14  1.

#### **Висновки**:
На заданій матриці рівень точності $\varepsilon = 10^{-4}$, при використанні QR-методу з використанням QR розкладу з бібліотеки ```scipy```, досягається за 28 ітерацій. Додатково перевіривши результати за допомогою ```numpy``` (результати якої вважитимемо за дійсні) можемо помітити, що похибка зосереджена в найбільших двох за модулем власних значеннях, різниця яких з результатами ```numpy``` має порядок $10^{-2}$, тоді як у інших не перевищує $\varepsilon$. Таким чином цей метод є найповільнишим з розглянутих, проте єдиним з них, що може працювати з несиметричними матрицями.

# Висновки
Під час виконання лабораторної роботи було на мові програмування ```Python``` з використанням бібліотеки матричних розрахунків ```numpy``` реалізовано три методи: метод скалярних добутків для знаходження старшого власного значення та вектора, та два методи розв'язання проблеми власних значень - метод Якобі та QR-метод. На мою думку, серед розглянутих, метод Якобі виявився найбільш швидким та точним при знаходженні власних векторів (при одноковій точності для власних чисел), проте він працює лише для симетричних матриць, та потребує багато розрахунків, отже його використання не завжди є доцільним. Також, слід зазначити, що окрім точності самих алгоритмів на результати вплинули точність компьютерного представлення чисел та відповідно компьютерних розрахунків.